kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/JSONInputFormat.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2013 Cloudera Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.kitesdk.data.spi.filesystem;
+
+import java.io.IOException;
+import java.util.List;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.JobContext;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
+import org.apache.hadoop.mapreduce.lib.input.FileSplit;
+import org.kitesdk.compat.Hadoop;
+import org.kitesdk.data.DatasetDescriptor;
+import org.kitesdk.data.spi.DataModelUtil;
+import org.kitesdk.data.spi.EntityAccessor;
+
+class JSONInputFormat<E> extends FileInputFormat<E, Void> {
+  private DatasetDescriptor descriptor = null;
+  private EntityAccessor<E> accessor = null;
+
+  public void setDescriptor(DatasetDescriptor descriptor) {
+    this.descriptor = descriptor;
+  }
+
+  public void setType(Class<E> type) {
+    this.accessor = DataModelUtil.accessor(type, descriptor.getSchema());
+  }
+
+  @Override
+  protected boolean isSplitable(JobContext context, Path filename) {
+    return false;
```

Comment:
Some JSON is splittable on newlines, but it is more complicated:
* "Whitespace can be inserted between any pair of tokens," so newlines aren't a universal record delimiter
* The parser doesn't yeild whitespace tokens, so implementing a newline rule (for opt-in cases that are safe) requires more work
* For encodings that support multi-byte characters, we need to ensure we can start at the next character to scan for `}{`

I also agree that format 3 should be supported, but as a format option. I'm also not sure that would be splittable at all, though it would be possible to concatenate long list object and split between those by looking for `][`.

> Splittability is very important for large datasets.

I agree, but it isn't necessarily a high priority. I'd like to get some basic support in and then assess the priority of splitting CSV and JSON. Right now we encourage a one-time import and haven't had anyone asking for splitting. We can also wrap dataset directories, but we don't currently expose a convenient way to set that up (a CLI command).

I've opened [CDK-880](https://issues.cloudera.org/browse/CDK-880) to make CSV splittable and [CDK-881](https://issues.cloudera.org/browse/CDK-881) to make JSON splittable. I don't think CSV splits are too hard, but let's follow up on that issue to discuss how to do them properly.